README.md
<!-- badges: start -->
[![CRAN
Version](https://img.shields.io/cran/v/bama?style=flat-square&color=blue&label=CRAN)](https://cran.r-project.org/package=bama)
[![GitHub
Release](https://img.shields.io/github/v/release/umich-cphds/bama?include_prereleases&label=Github&style=flat-square)](https://github.com/umich-cphds/bama)
[![Travis
CI](https://img.shields.io/travis/umich-cphds/bama?style=flat-square)](https://travis-ci.org/umich-cphds/bama)

Bayesian Mediation Analysis
===========================

Perform mediation analysis in the presence of high-dimensional mediators
based on the potential outcome framework. Bayesian Mediation Analysis
(BAMA), developed by Song et al (2019) , relies on two Bayesian sparse
linear mixed models to simultaneously analyze a relatively large number
of mediators for a continuous exposure and outcome assuming a small
number of mediators are truly active. This sparsity assumption also
allows the extension of univariate mediator analysis by casting the
identification of active mediators as a variable selection problem and
applying Bayesian methods with continuous shrinkage priors on the
effects.

Installation
------------

You can install `bama` via CRAN

    install.packages("bama")

Or devtools

    devtools::install_github("umich-cphds/bama", build_opts = c())

The Github version may contain new features or bug fixes not yet present
on CRAN, so if you are experiencing issues, you may want to try the
Github version of the package.

If you wish to install the package via `devtools`, you will need a C++
compiler installed. This can be accomplished by installing Rtools on
Windows and Xcode on MacOS.

Example
-------

This example is taken from the `bama` help file to help you get started
using the method. Please check the documentation of the function by
typing `?bama::bama`, and the vignette by typing `vingette("bama")` in
R.

`bama` includes an example dataset, `bama.data`. It is a `data.frame`
with a numeric response `y`, numeric exposure `a` and 100 numeric
mediators named `m1, m2, ..., m100`.

We recommend using much larger numbers for `burnin` and `ndraws`, for
example (30000, 1000).

    library(bama)

    Y <- bama.data$y
    A <- bama.data$a

    # grab the mediators from the example data.frame
    M <- as.matrix(bama.data[, paste0("m", 1:100)], nrow(bama.data))

    # We just include the intercept term in this example.
    C1 <- matrix(1, nrow(M), 1)
    C2 <- matrix(1, nrow(M), 1)
    beta.m  <- rep(0, 100)
    alpha.a <- rep(0, 100)

    set.seed(1245)
    bama.out <- bama(Y, A, M, beta.m, alpha.a, C1 = C2, C2 = C2, burnin = 1000,
                     ndraws = 100)

    # Rank mediators and see summary information
    head(summary(bama.out, rank = T))
    #>        estimate    ci.lower     ci.upper  pip
    #> m12  0.18576755  0.12218219  0.253780289 0.99
    #> m65 -0.23979113 -0.30642142 -0.165385815 0.99
    #> m89 -0.14252621 -0.21202100 -0.065313879 0.91
    #> m97 -0.04289325 -0.10325313  0.007526357 0.31
    #> m90 -0.03041407 -0.08301040  0.012826983 0.17
    #> m93  0.04051381 -0.00547444  0.118579967 0.17

Reference
=========

Song, Y, Zhou, X, Zhang, M, et al. Bayesian shrinkage estimation of high
dimensional causal mediation effects in omics studies. Biometrics. 2019;
1-11. [doi:10.1111/biom.13189](https://doi.org/10.1111/biom.13189)


NEWS.md
# bama v1.0.0

## New Features
* `bama` now contains `fdr.bama`, a `bama` variant that controls for the false
    discovery rate by estimating the null PIP distribution via the permutation
    test. It contains options to run in parallel (ie using the `parallel` package)
    as well, since the permutation test can be time consuming. `fdr.bama` also
    contains a summary function, `summary.fdr.bama` to help summarize results.

## Major Changes
* `bama` (and `fdr.bama`) now contain a weights parameter, `weights`.
* Change k, lm0, lm1, l from hard coded parameters in the C++ code to user
    set parameters (with defaults).
* Revert change that combined C1 and C2.

## Minor Changes
* Partial revamp `bama` documentation.
* Edit DESCRIPTION, and README / vignette introduction.
* Update DOI with (online) version of record.
* Redo `bama` type checking.

# bama v0.9.1

## Major Changes

* The most notable feature in this release is the renaming of `hdbm` to `bama`
    to prevent confusion with regards to its relationship with `hdmm`.
* bama now only accepts a single extra covariate matrix C, instead of C1 and C2.
## Minor Changes

* Added a summary function to `bama` to calculate the PIP, estimate, and
    credible interval of each mediator and optionally rank them.
* Added data quality check to ensure the matrix column norms in M, C1 and C2
    are not too small. If they are too small, `NaNs`
could be generated because of overflow problems when dividing by the norm.
* Improved documentation
* Added checks to C, the matrix of extra covariates, and alpha.a, the initial
alpha.a values.

# hdbm v0.9.0

* Initial version released on CRAN.


R/fdr.bama.R
#' Bayesian Mediation Analysis Controlling For False Discovery
#'
#' \code{fdr.bama} uses the permutation method to estimate the null PIP
#' distribution for each mediator and determines a threshold (based off of the
#' \code{fdr} parameter) for significance.
#'
#' TODO
#' @param Y Length \code{n} numeric outcome vector
#' @param A Length \code{n} numeric exposure vector
#' @param M \code{n x p} numeric matrix of mediators of Y and A
#' @param C1 \code{n x nc1} numeric matrix of extra covariates to include in the
#'     outcome model
#' @param C2 \code{n x nc2} numeric matrix of extra covariates to include in the
#'     mediator model
#' @param beta.m Length \code{p} numeric vector of initial \code{beta.m} in the
#'     outcome model
#' @param alpha.a Length \code{p} numeric vector of initial \code{alpha.a} in
#'     the mediator model
#' @param burnin Number of iterations to run the MCMC before sampling
#' @param ndraws Number of draws to take from MCMC after the burnin period
#' @param fdr False discovery rate. Default is 0.1
#' @param npermutations The number of permutations to generate while estimating
#'     the null pip distribution. Default is 200
#' @param k Shape parameter prior for inverse gamma. Default is 2.0
#' @param lm0 Scale parameter prior for inverse gamma for the small normal
#'    components. Default is 1e-4
#' @param lm1 Scale parameter prior for inverse gamma for the large normal
#'    components. Default is 1.0
#' @param l Scale parameter prior for the other inverse gamma distributions.
#'     Default is 1.0
#' @param mc.cores The number of cores to use while running \code{fdr.bama}.
#'     \code{fdr.bama} uses the \code{parallel} package for parallelization,
#'     so see that for more information. Default is 1 core
#' @param type Type of cluster to make when \code{mc.cores > 1}. See
#'     \code{makeCluster} in the \code{parallel} package for more details.
#'     Default is "PSOCK"
#' @return
#' \code{fdr.bama} returns a object of type "fdr.bama" with 6 elements:
#' \describe{
#' \item{bama.out}{Output from the \code{bama} run.}
#' \item{pip.null}{A \code{p x npermutations} matrices containing the
#'     estimated null PIP distribution for each mediator.
#' }
#' \item{threshold}{The cutoff significance threshold for each PIP controlling
#'     for the false discovery rate.
#' }
#' \item{fdr}{The false discovery rate used to calculate \code{threshold}.}
#' \item{call}{The R call that generated the output.}
#' }
#' @examples
#' library(bama)
#'
#' Y <- bama.data$y
#' A <- bama.data$a
#'
#' # grab the mediators from the example data.frame
#' M <- as.matrix(bama.data[, paste0("m", 1:100)], nrow(bama.data))
#'
#' # We just include the intercept term in this example as we have no covariates
#' C1 <- matrix(1, 1000, 1)
#' C2 <- matrix(1, 1000, 1)
#' beta.m  <- rep(0, 100)
#' alpha.a <- rep(0, 100)
#'
#' set.seed(12345)
#' \donttest{
#' out <- fdr.bama(Y, A, M, C1, C2, beta.m, alpha.a, burnin = 1000,
#'                 ndraws = 100, npermutations = 10)
#'
#' # The package includes a function to summarise output from 'fdr.bama'
#' summary(out)
#' }
#' @references
#' Song, Y, Zhou, X, Zhang, M, et al. Bayesian shrinkage estimation of high
#' dimensional causal mediation effects in omics studies. Biometrics. 2019;
#' 1-11. \href{http://doi.org/10.1111/biom.13189}{doi:10.1111/biom.13189}
#' @author Alexander Rix
#' @export
fdr.bama <- function(Y, A, M, C1, C2, beta.m, alpha.a, burnin, ndraws,
                     npermutations = 200, fdr = 0.1, k = 2.0, lm0 = 1e-4,
                     lm1 = 1.0, l = 1.0, mc.cores = 1, type = "PSOCK")
{
    call <- match.call()

    bama.out <- bama(Y, A, M, C1, C2, beta.m, alpha.a, burnin, ndraws, k,
                     lm0, lm1, l)

    if (npermutations <= 0)
        stop("'npermutations' must be a positive integer.")

    if (fdr <= 0 || fdr >=1)
        stop("'fdr' must be in the interval (0, 1).")

    pi2 <- colMeans(bama.out$r1 == 0 & bama.out$r3 == 1)
    pi3 <- colMeans(bama.out$r1 == 1 & bama.out$r3 == 0)
    pi4 <- colMeans(bama.out$r1 == 0 & bama.out$r3 == 0)

    seeds <- sample(.Machine$integer.max, npermutations + 1)
    permute.bama <- function(i)
    {
        set.seed(seeds[i])

        n <- length(Y)
        bama.r1 <- run_bama_mcmc(Y[sample(n)], A, M, C1, C2, beta.m, alpha.a,
                                 burnin, ndraws, k, lm0, lm1, l)

        bama.r3 <- run_bama_mcmc(Y, A, M[sample(n), ], C1, C2, beta.m, alpha.a,
                                 burnin, ndraws, k, lm0, lm1, l)

        bama.r1.r3 <- run_bama_mcmc(Y[sample(n)], A, M[sample(n), ], C1, C2,
                                    beta.m, alpha.a, burnin, ndraws, k, lm0,
                                    lm1, l)

        p2 <- colMeans(bama.r1$r1 == 0 & bama.r1$r3 == 1)
        p3 <- colMeans(bama.r3$r1 == 1 & bama.r3$r3 == 0)
        p4 <- colMeans(bama.r1.r3$r1 == 0 & bama.r1.r3$r3 == 0)

        pi2 / (p2 + p3 + p4) * p2 + pi3 / (p2 + p3 + p4) * p3 +
        pi4 / (p2 + p3 + p4) * p4
    }

    if (mc.cores == 1) {
        pip.null <- sapply(seq(npermutations), permute.bama)
    }
    else {
        cl <- parallel::makeCluster(mc.cores, type = type)
        parallel::clusterExport(cl, list("Y", "A", "M", "C1", "C2", "beta.m",
                                         "alpha.a", "burnin", "ndraws", "k",
                                         "lm0", "lm1", "l", "pi2",  "pi3",
                                         "pi4", "seeds"),
                                envir = environment()
        )

        pip.null <- parallel::parSapply(cl, seq(npermutations), permute.bama)
        parallel::stopCluster(cl)
    }
    set.seed(seeds[npermutations + 1])

    threshold <- apply(pip.null, 1, stats::quantile, probs = 1 - fdr)
    names(threshold) <- colnames(M)

    structure(list(bama.out = bama.out, pip.null = pip.null,
                   threshold = threshold, fdr = fdr, call = call),
              class = "fdr.bama")
}

#' Summarize objects of type "fdr.bama"
#'
#' \code{summary.fdr.bama} summarizes the \code{beta.m} estimates from
#' \code{fdr.bama} and for each mediator generates an overall estimate,
#' credible interval, posterior inclusion probability (PIP), and PIP threshold
#' for significance controlling for the specified false discovery rate (FDR).
#' @return A data.frame with 4 elements. The beta.m estimates, the estimates'
#'     *credible* interval (which by default is 95\%), and the posterior
#'      inclusion probability (pip) of each 'beta.m'.
#' @param object An object of class "bama".
#' @param rank Whether or not to rank the output by posterior inclusion
#'     probability. Default is TRUE.
#' @param ci The credible interval to calculate. \code{ci} should be a length 2
#'     numeric vector specifying the upper and lower bounds of the CI. By
#'     default, \code{ci = c(0.025, .975)}.
#' @param fdr False discovery rate. By default, it is set to whatever the
#' \code{fdr} of \code{object} is. However, it can be changed to recalculate
#' the PIP cutoff threshold.
#' @param filter Whether or not to filter out mediators with PIP less than the
#'     PIP threshold.
#' @param ... Additional optional arguments to \code{summary}
#' @export
summary.fdr.bama <- function(object, rank = F, ci = c(0.025, 0.975),
                             fdr = object$fdr, filter = T, ...)
{
    if (class(object) != "fdr.bama")
        stop("'object' is not an bama object.")

    if (!is.logical(rank) || length(rank) != 1)
    stop("'rank' should be a length 1 logical.")

    if (!is.numeric(ci) || length(ci) != 2)
        stop("'ci' should be a length 2 numeric.")
    if (ci[1] >= ci[2])
        stop("'ci[1]' should be less than 'ci[2]'.")

    pip    <- colMeans(object$bama.out$r1 * object$bama.out$r3)
    beta.m <- colMeans(object$bama.out$beta.m)

    ci.l <- apply(object$bama.out$beta.m, 2, stats::quantile, probs = ci[1])
    ci.h <- apply(object$bama.out$beta.m, 2, stats::quantile, probs = ci[2])

    out <- data.frame(estimate = beta.m, ci.lower = ci.l,
                           ci.upper = ci.h, pip = pip,
                           pip.threshold = object$threshold)
    if (rank)
        out <- out[order(pip, decreasing = T), ]

    if (filter)
        out <- out[which(out$pip > out$pip.threshold), ]

    out
}

#' Printing bama objects
#'
#' Print a bama object.
#' @param x An object of class 'bama'.
#' @param ... Additional arguments to pass to print.data.frame or summary.bama
#' @export
print.fdr.bama <- function(x , ...)
{
    print(summary(x, ...), ...)
}
